### PR TITLE
Add support for env var kind

### PR DESCRIPTION
**Pull request by lcarva**
* CLOUDBLD-1484

### CODE REVIEW

File: cachito/web/models.py [hunk 1]
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
...
+        :param str name: the name of the environment variable
+        :param dict info: the description of the environment variable. Must include "value" and
+            "kind" attributes
         :raises ValidationError: if the environment variable is invalid
         """
-        if not isinstance(value, str):
+        if not isinstance(name, str):
+            raise ValidationError("The name of environment variables must be a string")
+        if not isinstance(info, dict):
+            raise ValidationError("The info of environment variables must be an object")
```
Comment (nirzari) [inline cachito/web/models.py:705]: Do we want to specify `a dict object` instead ?

Comment (lcarva) [reply cachito/web/models.py:705]: It's a little awkward because this error bubbles up to the API error response which deals in term of Json, not python.

File: README.md [hunk 2]
```diff
@@ -190,7 +190,9 @@ Custom configuration for the Celery workers are listed below:
   writeable.
 * `cachito_default_environment_variables` - a dictionary where the keys are names of package
   managers. The values are dictionaries where the keys are default environment variables to
-  set for that package manager and the values are the environment variable values. Check
+  set for that package manager and the values a dictionary with the keys `value` and `kind`. The
+  `value` must be a string which specifies the value of the environment variable. The `kind` must
+  also be a string which specifies the type of value, either `"path"`, or `"string"`. Check
```
Comment (mprahl) [inline README.md:195]:
```suggestion
  also be a string which specifies the type of value, either `"path"` or `"string"`. Check
```

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+"""Add environment variables types
+
+Revision ID: cb6bdbe533cc
+Revises: 615c19a1cee1
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:4]: FYI, one of us will have to modify this depending on if this PR is merged before #181.

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+# revision identifiers, used by Alembic.
+revision = "cb6bdbe533cc"
+down_revision = "b46cf36806d7"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:21]:
```suggestion
    with op.batch_alter_table("environment_variable") as batch_op:
```

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        # Drop previous unique_constraint. Since the constraint was not originally added with a
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:22]: Can you just look at the automatic name used by Postgres and fix the old migration to set the name explicitly instead?

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        if old_unique_constraint:
+            batch_op.drop_constraint(old_unique_constraint["name"], type_="unique")
+
+        # Make this nullable initially, so we can adjust the data first
+        batch_op.add_column(sa.Column("kind", sa.String(), nullable=True))
+
+    connection = op.get_bind()
+    connection.execute("UPDATE environment_variable SET kind='path' WHERE kind IS NULL")
+
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:41]:
```suggestion
    with op.batch_alter_table("environment_variable") as batch_op:
```

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                old_unique_constraint = constraint
+                break
+        if old_unique_constraint:
+            batch_op.drop_constraint(old_unique_constraint["name"], type_="unique")
+
+        # Make this nullable initially, so we can adjust the data first
+        batch_op.add_column(sa.Column("kind", sa.String(), nullable=True))
+
+    connection = op.get_bind()
+    connection.execute("UPDATE environment_variable SET kind='path' WHERE kind IS NULL")
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:39]: Could you please use SQLAlchemy to do this and not rely on raw SQL?

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
+        batch_op.create_unique_constraint(
+            "uq_environmet_variable_name_value_kind", ["name", "value", "kind"]
+        )
+        batch_op.alter_column("kind", nullable=False)
+
+
+def downgrade():
+    with op.batch_alter_table("environment_variable", schema=None) as batch_op:
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:49]:
```suggestion
    with op.batch_alter_table("environment_variable") as batch_op:
```

File: cachito/web/models.py [hunk 9]
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
...
+
+        required_keys = {"value", "kind"}
+        missing_keys = required_keys - info.keys()
+        if missing_keys:
+            raise ValidationError(
+                "The following attributes must be set in the info of the environment variables: "
+                f"{', '.join(sorted(missing_keys))}"
```
Comment (mprahl) [inline cachito/web/models.py:712]:
```suggestion
                ", ".join(sorted(missing_keys))
```

File: cachito/web/models.py [hunk 10]
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
...
+            )
+
+        invalid_keys = info.keys() - required_keys
+        if invalid_keys:
+            raise ValidationError(
+                "The following attributes are not allowed in the info of the environment "
```
Comment (mprahl) [inline cachito/web/models.py:718]:
```suggestion
                "The following keys are not allowed in the info of the environment "
```

File: cachito/web/models.py [hunk 11]
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
...
+                f"variables: {', '.join(sorted(invalid_keys))}"
+            )
+
+        if not isinstance(info["value"], str):
             raise ValidationError("The value of environment variables must be a string")
+        kind = info.get("kind")
+        if not isinstance(kind, str):
+            raise ValidationError("The kind of environment variables must be a string")
+        if kind not in ("path", "string"):
```
Comment (mprahl) [inline cachito/web/models.py:727]: Optional: It'd be nice to define the valid kinds at the class level.

File: cachito/web/models.py [hunk 12]
```diff
@@ -680,42 +685,69 @@ class EnvironmentVariable(db.Model):
...
+        :param dict info: the description of the environment variable
         :return: the EnvironmentVariable object
         :rtype: EnvironmentVariable
         """
-        cls.validate_json(name, value)
-        return cls(name=name, value=value)
+        cls.validate_json(name, info)
+        return cls(name=name, **info)
 
     def to_json(self):
```
Comment (mprahl) [inline cachito/web/models.py:743]: I don't think this method is used anymore. Could you please delete it if that's the case?

File: cachito/web/models.py (hunk 11 above)
Comment (mprahl) [inline cachito/web/models.py:727]: Instead of the kind of `string`, how about `literal`?

File: cachito/web/static/api_v1.yaml [hunk 13]
```diff
@@ -360,6 +397,21 @@ components:
...
+      properties:
+        value:
+          type: string
+          description: The value of the environment variable
+        kind:
+          type: string
+          description: The kind of the environment variable, e.g. "path" or "string"
+      required:
+      - value
+      - kind
```
Comment (mprahl) [inline cachito/web/static/api_v1.yaml:411]: Could you alphabetize this please?

File: cachito/workers/config.py [hunk 14]
```diff
@@ -181,12 +184,23 @@ def validate_celery_config(conf, **kwargs):
...
+        if not isinstance(pkg_manager_value, dict):
             raise ConfigError(
                 'The configuration "cachito_default_environment_variables" must be a '
                 "dictionary of dictionaries"
             )
+        for env_var_info in pkg_manager_value.values():
+            if not isinstance(env_var_info, dict):
+                raise ConfigError(
+                    'The configuration "cachito_default_environment_variables" must be a '
+                    "dictionary of dictionaries of dictionaries!"
```
Comment (mprahl) [inline cachito/workers/config.py:197]: Say that five times fast!

File: cachito/workers/pkg_managers/general.py [hunk 15]
```diff
@@ -111,7 +111,9 @@ def update_request_with_packages(request_id, packages, env_vars=None):
     Update the request with the resolved packages and corresponding metadata.
 
     :param list packages: the list of packages that were resolved
-    :param dict env_vars: mapping of environment variables to record
+    :param dict env_vars: mapping of environment variables to record. Attributes represent
```
Comment (mprahl) [inline cachito/workers/pkg_managers/general.py:114]:
```suggestion
    :param dict env_vars: mapping of environment variables to record. The keys represent
```

File: cachito/workers/tasks/gomod.py [hunk 16]
```diff
@@ -43,7 +43,10 @@ def fetch_gomod_source(request_id, dep_replacements=None):
         log.exception("Failed to fetch gomod dependencies for request %d", request_id)
         raise
 
-    env_vars = {"GOCACHE": "deps/gomod", "GOPATH": "deps/gomod"}
-    env_vars.update(config.cachito_default_environment_variables.get("gomod", {}))
+    env_vars = {
+        "GOCACHE": {"value": "deps/gomod", "kind": "path"},
+        "GOPATH": {"value": "deps/gomod", "kind": "path"},
+    }
+    env_vars.update(get_worker_config().cachito_default_environment_variables.get("gomod", {}))
```
Comment (mprahl) [inline cachito/workers/tasks/gomod.py:50]:
```suggestion
    env_vars.update(config.cachito_default_environment_variables.get("gomod", {}))
```

File: tests/test_api_v1.py [hunk 17]
```diff
@@ -1172,3 +1196,54 @@ def test_request_config_post_not_authorized(auth_env, client, db):
...
+    if env_vars is not None:
+        payload = {"environment_variables": env_vars}
+        patch_rv = client.patch(
+            f"/api/v1/requests/{request.id}", json=payload, environ_base=worker_auth_env
+        )
+        assert patch_rv.status_code == 200
+
+    env_vars_expected = env_vars or {}
+
+    len(EnvironmentVariable.query.all()) == len(env_vars_expected.items())
```
Comment (mprahl) [inline tests/test_api_v1.py:1232]: This may work:
```suggestion
    EnvironmentVariable.query.count() == len(env_vars_expected.items())
```

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py (hunk 5 above)
Comment (lcarva) [reply cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:22]: IMO that's too brittle. We should instead setup explicit [naming conventions](https://alembic.sqlalchemy.org/en/latest/naming.html), and create another migration script to "rename" them all.

File: cachito/web/models.py (hunk 10 above)
Comment (lcarva) [reply cachito/web/models.py:718]: This has to say attributes because it bubbles up to the API which is JSON based, no?

File: cachito/web/models.py (hunk 9 above)
Comment (lcarva) [reply cachito/web/models.py:712]: This doesn't work. It first concatenates the string, then calls join on it.

File: README.md [hunk 18]
```diff
@@ -190,7 +190,9 @@ Custom configuration for the Celery workers are listed below:
   writeable.
 * `cachito_default_environment_variables` - a dictionary where the keys are names of package
   managers. The values are dictionaries where the keys are default environment variables to
-  set for that package manager and the values are the environment variable values. Check
+  set for that package manager and the values a dictionary with the keys `value` and `kind`. The
```
[the remark this answers is not in the document]
Comment (lcarva) [reply README.md]: I went with "and the values are a dictionary"

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py (hunk 3 above)
Comment (lcarva) [reply cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:4]: Ack.

File: cachito/web/models.py (hunk 11 above)
Comment (lcarva) [reply cachito/web/models.py]: Good idea. An environment variable must always be a string. Path and literal are just different "types" of strings. Made the change.

File: cachito/workers/config.py (hunk 14 above)
Comment (lcarva) [reply cachito/workers/config.py:197]: I did and my dictionary jumped off the shelf.

File: tests/test_api_v1.py (hunk 17 above)
Comment (lcarva) [reply tests/test_api_v1.py]: It did!

**Comment by lcarva** [comment]
I've addressed all the comments. This is ready for another review round.

File: cachito/web/models.py (hunk 10 above)
Comment (mprahl) [reply cachito/web/models.py:720]: I realize that, but we use the term `keys` everywhere else as far as I know for user facing errors.

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py (hunk 5 above)
Comment (mprahl) [reply cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:32]: We have successfully done this in the past:
* https://github.com/release-engineering/cachito/commit/3bfee393276abc0d5b0c59f18b0340d3129d2689
* https://github.com/release-engineering/cachito/commit/dbb205747d089cfadd4418741a5aa09a1942d778

File: tests/test_api_v1.py [hunk 19]
```diff
@@ -1172,3 +1196,54 @@ def test_request_config_post_not_authorized(auth_env, client, db):
...
+    if env_vars is not None:
+        payload = {"environment_variables": env_vars}
+        patch_rv = client.patch(
+            f"/api/v1/requests/{request.id}", json=payload, environ_base=worker_auth_env
+        )
+        assert patch_rv.status_code == 200
+
+    env_vars_expected = env_vars or {}
+
+    EnvironmentVariable.query.count() == len(env_vars_expected.items())
```
Comment (mprahl) [inline tests/test_api_v1.py:1232]:
```suggestion
    EnvironmentVariable.query.count() == len(env_vars_expected)
```

File: cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py [hunk 20]
```diff
@@ -23,7 +23,7 @@ def upgrade():
         sa.Column("name", sa.String(), nullable=False),
         sa.Column("value", sa.String(), nullable=False),
         sa.PrimaryKeyConstraint("id"),
-        sa.UniqueConstraint("name", "value"),
+        sa.UniqueConstraint("name", "value", name="uq_environmet_variable_name_value"),
```
Comment (mprahl) [inline cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py:26]: @lcarva do you mind using the value that Postgres generates by default since we've been using that for other constraints and it'd be nice to be consistent.

Comment (lcarva) [reply cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py:26]: Sure. What would that be here?

Comment (mprahl) [reply cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py:26]:
```suggestion
        sa.UniqueConstraint("name", "value", name="environment_variable_name_value_key"),
```

Comment (lcarva) [reply cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py]: Done!

**Comment by lcarva** [comment]
Rebased, updated the migration script to use the newly added migration script as the `down_revision`.

I also caught a bug in the migration script:
```diff
-        env_var_table.update().where(env_var_table.c.kind is None).values(kind="path")
+        env_var_table.update().where(env_var_table.c.kind == sa.null()).values(kind="path")
```

File: cachito/web/models.py (hunk 10 above)
Comment (lcarva) [reply cachito/web/models.py]: Alright, I changed this to be consistent with the other instances. 

BTW, we should also change `The following parameters are invalid` to be consistent with this. I can do that in a follow up PR.

File: cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py [hunk 21]
```diff
@@ -23,7 +23,7 @@ def upgrade():
         sa.Column("name", sa.String(), nullable=False),
         sa.Column("value", sa.String(), nullable=False),
         sa.PrimaryKeyConstraint("id"),
-        sa.UniqueConstraint("name", "value"),
+        sa.UniqueConstraint("name", "value", name="environment_variable_name_value_key"),
```
Comment (mprahl) [inline cachito/web/migrations/versions/cdf17fad3edb_request_env_vars.py:26]: Optional: This change would be better as a separate commit

File: cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+    sa.Column("id", sa.Integer(), primary_key=True),
+    sa.Column("name", sa.String, nullable=False),
+    sa.Column("value", sa.String, nullable=False),
+    sa.Column("kind", sa.String, nullable=False),
+)
+
+
+def upgrade():
+    with op.batch_alter_table("environment_variable") as batch_op:
+        # Drop previous unique_constraint. Since the constraint was not originally added with a
```
Comment (mprahl) [inline cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py:31]: Please delete this comment

Comment (lcarva) [reply cachito/web/migrations/versions/cb6bdbe533cc_add_env_var_types.py]: Doh! removed it.
